examples/utils/example-utils/src/migrationTool/sameContainerMigrationTool.ts

```diff
@@ -0,0 +1,460 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IPactMap, PactMap } from "@fluid-experimental/pact-map";
+import { DataObject, DataObjectFactory } from "@fluidframework/aqueduct";
+import type { IFluidHandle } from "@fluidframework/core-interfaces";
+import { ISequencedDocumentMessage, MessageType } from "@fluidframework/protocol-definitions";
+
+import type { ISameContainerMigrationTool } from "../migrationInterfaces";
+
+const pactMapKey = "pact-map";
+const newVersionKey = "newVersion";
+
+// RANDOM NOTES
+// Should the migration tool emit state changes and such REGARDLESS of connection state, etc.?
+// It would be simpler to just observe the state changing and report that, but it externalizes the knowledge that we might still be anticipating the state to continue
+// changing as we connect and that the Migrator should NOT take action.  Otherwise the Migrator would need to have the knowledge that it shouldn't immediately act upon
+// the state changes if not connected.
+
+export class SameContainerMigrationTool extends DataObject implements ISameContainerMigrationTool {
+	private _pactMap: IPactMap<string> | undefined;
+
+	/**
+	 * A promise that is only defined if the local client has made a proposal, and will resolve when any proposal goes pending.
+	 * Not all clients will have a _proposalP, if they do not make the proposal.  Those clients instead will only see a remote proposal come in and advance
+	 * directly from "collaborating" to "stoppingCollaboration".
+	 */
+	private _proposalP: Promise<void> | undefined;
+	/**
+	 * A promise that will resolve when the proposal is either pending or accepted, signalling that we have moved on to a later stage of the migration.
+	 */
+	private _pendingP: Promise<void> | undefined;
+	/**
+	 * A promise that will resolve when the proposal is accepted.
+	 */
+	private _acceptedP: Promise<void> | undefined;
+	/**
+	 * A promise that will resolve when we know the final v1 summary was successfully submitted.
+	 * Note that even when loading from that summary, we should expect to see the summaryAck as part of the logTail
+	 */
+	private _v1SummaryP: Promise<void> | undefined;
+	/**
+	 * This boolean will make it easier to synchronously determine if the v1 summary is done.
+	 */
+	private _v1SummaryDone: boolean = false;
```

Comment:
nit: I wonder if there is a more elegant way to do this? I'm interested since we use this pattern a few times in this file

Reply:
There might be, but a challenge here is that we have several places that need to observe current state synchronously (in particular promise executors and/or event handlers which should always be synchronous).  Open to ideas, but the requirements may shift a little as we develop the flow so might be more convenient to explore better options a little further on.